Encoding issues under Windows

I'm running into a crash related to encoding on Windows:

```
(.venv) C:\Users\Brecht\My Documents\Opqode\rinohtype>pytest tests_regression -k test_version[inprocess]
============================= test session starts =============================
platform win32 -- Python 3.8.6, pytest-5.4.3, py-1.8.2, pluggy-0.13.1
rootdir: C:\Users\Brecht\My Documents\Opqode\rinohtype, inifile: setup.cfg
plugins: assume-2.2.1, console-scripts-0.2.0
collected 41 items / 40 deselected / 1 selected

..\..\..\My Documents\Opqode\rinohtype\test_rinoh.py F                   [100%]

================================== FAILURES ===================================
___________________________ test_version[inprocess] ___________________________

script_runner = <ScriptRunner inprocess>

    def test_version(script_runner):
        ret = script_runner.run('rinoh', '--version')
>       assert ret.success
E       assert False
E        +  where False = <pytest_console_scripts.RunResult object at 0x000001F1725F7520>.success

C:\Users\Brecht\Documents\Opqode\rinohtype\tests_regression\test_rinoh.py:21: AssertionError
---------------------------- Captured stdout call -----------------------------
# Running console script: rinoh --version
# Script return code: 1
# Script stdout:

# Script stderr:
Traceback (most recent call last):
  File "c:\users\brecht\my documents\opqode\rinohtype\.venv\lib\site-packages\py\_path\common.py", line 177, in read
    return f.read()
  File "C:\Python38\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 2: character maps to <undefined>

=========================== short test summary info ===========================
FAILED ..\..\..\My Documents\Opqode\rinohtype\test_rinoh.py::test_version[inprocess]
====================== 1 failed, 40 deselected in 1.73s =======================
```
Running `rinoh --version` in the same terminal runs without issues.

I can't seem to figure out how to get a full traceback pointing out the relevant piece of pytest-console-scripts code.

Thanks, Vasily. It looks like your guess is correct. This is the relevant part of the traceback wih the `raise` in place:

```
script_runner = <ScriptRunner inprocess>

    def test_version(script_runner):
>       ret = script_runner.run('rinoh', '--version')

tests_regression\test_rinoh.py:20:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ScriptRunner inprocess>, command = 'rinoh', arguments = ('--version',), options = {}

    def run(self, command, *arguments, **options):
        print('# Running console script:', command, *arguments)
        if self.launch_mode == 'inprocess':
>           return self.run_inprocess(command, *arguments, **options)

..\..\code\pytest-console-scripts\pytest_console_scripts.py:126:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ScriptRunner inprocess>, command = 'rinoh', arguments = ('--version',), options = {}
cmdargs = ['rinoh', '--version']
script = local('C:\\Users\\Brecht\\Documents\\Opqode\\rinohtype\\.venv\\Scripts\\rinoh.EXE')
stdin = <_io.StringIO object at 0x000001D95B3FF310>, stdout = <_io.StringIO object at 0x000001D95B3FF5E0>
stderr = <_io.StringIO object at 0x000001D95B3FF670>, returncode = 0
stdin_patch = <mock.mock._patch object at 0x000001D95B3EE100>
stdout_patch = <mock.mock._patch object at 0x000001D95B3EE430>
stderr_patch = <mock.mock._patch object at 0x000001D95B3EE0D0>
argv_patch = <mock.mock._patch object at 0x000001D95B3EE0A0>

    def run_inprocess(self, command, *arguments, **options):
        cmdargs = [command] + list(arguments)
        script = py.path.local(self._locate_script(command, **options))
        stdin = options.get('stdin', StreamMock())
        stdout = StreamMock()
        stderr = StreamMock()
        returncode = 0
        stdin_patch = mock.patch('sys.stdin', new=stdin)
        stdout_patch = mock.patch('sys.stdout', new=stdout)
        stderr_patch = mock.patch('sys.stderr', new=stderr)
        argv_patch = mock.patch('sys.argv', new=cmdargs)
        saved_dir = os.getcwd()
        logger_conf = self._save_and_reset_logger()

        if 'env' in options:
            old_env = os.environ
            os.environ = options.get('env')

        if 'cwd' in options:
            os.chdir(options['cwd'])
        with stdin_patch, stdout_patch, stderr_patch, argv_patch:
            try:
>               compiled = compile(script.read(), str(script), 'exec', flags=0)

..\..\code\pytest-console-scripts\pytest_console_scripts.py:194:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = local('C:\\Users\\Brecht\\Documents\\Opqode\\rinohtype\\.venv\\Scripts\\rinoh.EXE'), mode = 'r'

    def read(self, mode='r'):
        """ read and return a bytestring from reading the path. """
        with self.open(mode) as f:
>           return f.read()

..\..\..\my documents\opqode\rinohtype\.venv\lib\site-packages\py\_path\common.py:177:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <encodings.cp1252.IncrementalDecoder object at 0x000001D95B3EEB50>
input = b'MZ\x90\x00\x03\x00\x00\x00\x04\x00\x00\x00\xff\xff\x00\x00\xb8\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\...\x0b\x01\x00\xb7\xe3\x00\x00\xd0\xe3\x00\x004\x0f\x01\x00\xd0\xe3\x00\x00\x1c\xe4\x00\x000\x10\x01\x00\x00\x00\x00\x00'
final = True

    def decode(self, input, final=False):
>       return codecs.charmap_decode(input,self.errors,decoding_table)[0]
E       UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 2: character maps to <undefined>

C:\Python38\lib\encodings\cp1252.py:23: UnicodeDecodeError
------------------------------------------------ Captured stdout call -------------------------------------------------
# Running console script: rinoh --version
=============================================== short test summary info ===============================================
FAILED tests_regression/test_rinoh.py::test_version[inprocess] - UnicodeDecodeError: 'charmap' codec can't decode byt...
========================================== 1 failed, 40 deselected in 3.91s ===========================================
```

--- a/pytest_console_scripts.py
+++ b/pytest_console_scripts.py
@@ -3,13 +3,13 @@ from __future__ import unicode_literals, print_function
 import io
 import logging
 import os
+import pkg_resources
 import shutil
 import subprocess
 import sys
 import traceback
 
 import mock
-import py.path
 import pytest
 
 if sys.version_info.major == 2:
@@ -169,13 +169,29 @@ class ScriptRunner(object):
 
         raise FileNotFoundError('Cannot find ' + command)
 
+    def _load_script(self, command, **options):
+        """Load target script via entry points or compile/exec."""
+        entry_points = list(pkg_resources.iter_entry_points('console_scripts',
+                                                            command))
+        if entry_points:
+            return entry_points[0].load()
+
+        script_path = self._locate_script(command, **options)
+
+        def exec_script():
+            with open(script_path, 'rt', encoding='utf-8') as script:
+                compiled = compile(script.read(), str(script), 'exec', flags=0)
+                exec(compiled, {'__name__': '__main__'})
+            return 0
+
+        return exec_script
+
     def run_inprocess(self, command, *arguments, **options):
         cmdargs = [command] + list(arguments)
-        script = py.path.local(self._locate_script(command, **options))
+        script = self._load_script(command, **options)
         stdin = options.get('stdin', StreamMock())
         stdout = StreamMock()
         stderr = StreamMock()
-        returncode = 0
         stdin_patch = mock.patch('sys.stdin', new=stdin)
         stdout_patch = mock.patch('sys.stdout', new=stdout)
         stderr_patch = mock.patch('sys.stderr', new=stderr)
@@ -191,8 +207,7 @@ class ScriptRunner(object):
             os.chdir(options['cwd'])
         with stdin_patch, stdout_patch, stderr_patch, argv_patch:
             try:
-                compiled = compile(script.read(), str(script), 'exec', flags=0)
-                exec(compiled, {'__name__': '__main__'})
+                returncode = script()
             except SystemExit as exc:
                 returncode = exc.code
                 if isinstance(returncode, str):

--- a/pytest_console_scripts.py
+++ b/pytest_console_scripts.py
@@ -208,6 +208,8 @@ class ScriptRunner(object):
         with stdin_patch, stdout_patch, stderr_patch, argv_patch:
             try:
                 returncode = script()
+                if returncode is None:
+                    returncode = 0  # None also means success.
             except SystemExit as exc:
                 returncode = exc.code
                 if isinstance(returncode, str):

--- a/tests/test_run_scripts.py
+++ b/tests/test_run_scripts.py
@@ -76,6 +76,23 @@ def test_script(script_runner):
     assert result.success
 
 
+@pytest.mark.script_launch_mode('inprocess')
+def test_return_None(script_runner):
+    """Check that entry point function returning None is counted as success."""
+
+    # Many console_scripts entry point functions return 0 on success but not
+    # all of them do. Returning `None` is also allowed and would be translated
+    # to return code 0 when run normally via wrapper. This test checks that we
+    # handle this case properly in inprocess mode.
+    #
+    # One commonly available script that returns None from the entry point
+    # function is easy_install so we use it here.
+
+    result = script_runner.run('easy_install', '-h')
+    assert result.success
+    assert '--verbose' in result.stdout
+
+
 @pytest.mark.script_launch_mode('both')
 def test_abnormal_exit(console_script, script_runner):
     console_script.write('import sys;sys.exit("boom")')

